### PR TITLE
fix: #16 Add element ids in big bank example

### LINKED ISSUE #16: Partial Big Bank Example
Implement the majority of the big bank use cases

### PR DESCRIPTION
**Pull request by yt-ms**
* [X] fix Midnighter/structurizr-python#16
* [X] description of feature/fix
* [X] tests added/passed
* [ ] add an entry to the [next release](../CHANGELOG.rst)

Add IDs to the `big_bank.py` example, based on https://github.com/structurizr/dsl/blob/master/examples/big-bank-plc.dsl.  As part of this, extended `SoftwareSystem.add_container` to allow additional kwargs through, and made the technology optional (but still explicit).

Didn't think significant enough to add to the changelog.

THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE Apache License v.2.0. YOU MAY OBTAIN A COPY OF THE LICENSE AT https://www.apache.org/licenses/LICENSE-2.0.

THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: examples/big_bank.py
Comment (Midnighter) [inline examples/big_bank.py]: Please don't rename this function. I rely on the existence of a `main` function in every example to create the workspace in some other code. (Should be documented, of course.)

You're welcome to create other functions, though, called by `main` that construct a specific part of the example, as long as `main` returns the workspace.

File: src/structurizr/model/software_system.py
Comment (Midnighter) [inline src/structurizr/model/software_system.py:81]: :+1:

**Review by Midnighter — APPROVED** [review]
Changes look mostly good to me but I'd let @ilaif have the final word here.

One hint for the future: It'd be cleaner if you made one commit for the change to the software system and its tests and another commit for changing the example.

**Comment by yt-ms** [comment]
> One hint for the future: It'd be cleaner if you made one commit for the change to the software system and its tests and another commit for changing the example.

Ack

File: examples/big_bank.py
Comment (yt-ms) [reply examples/big_bank.py]: Done.

File: examples/big_bank.py
Comment (ilaif) [inline examples/big_bank.py:20]: So actually, we based the example on: https://github.com/structurizr/dotnet/blob/master/Structurizr.Examples/BigBankPlc.cs

And there, there's no mention of ids. I'm not event sure why we would need them if "name" is a unique identifier?

I don't know the DSL that well, but it seems that there's no real usage of the id, as the name is unique as well.

Can you elaborate more on this?

Comment (yt-ms) [reply examples/big_bank.py:20]: Sure.  One of the things I'm working on is "architecture as code", including integrating architecture change into our other change management processes across the enterprise.  For a given system, I want to be able to have a workspae that combines together high-level structure (Systems, Containers) which would be either defined through the Python API or in Structurizr itself, along with lower-level (Component, Code) which would be extracted from source code directly through the [Component Finder](https://github.com/structurizr/java-extensions/blob/master/docs/component-finder.md), [annotations](https://github.com/structurizr/java-extensions/blob/master/docs/structurizr-annotations.md) and so on.

Ultimately I want to have a file which can be managed in Git (and so have PR approval workflows) that represents the whole model, so the two obvious choices are JSON and the Structurizr DSL.  The JSON form isn't really designed for humans, and if you look at a typical diff then it's pretty hard to tell what's going on, but with the DSL then it's clear what is changing.  As long as you have IDs on things, else the Relationships become pretty opaque.  

Here's an example from BigBank, using a rudimentary DSL translator that I knocked together:
```
      1 -> 2 Uses
      2 -> 4 Uses
      2 -> 6 "Sends e-mail using"
      6 -> 1 "Sends e-mails to"
      9 -> 4 Uses
      1 -> 9 "Withdraws cash using"
```
and here's what it looks like with my proposed changes:
```
      customer -> internetBankingSystem "Views account balances, and makes payments using"
      internetBankingSystem -> mainframe "Gets account information from, and makes payments using"
      internetBankingSystem -> email "Sends e-mail using"
      email -> customer "Sends e-mails to"
      atm -> mainframe Uses
      customer -> atm "Withdraws cash using"
```
That second version is way more understandable in a diff than the first, which is why I'd like to support it.  I'm surprised that the .cs version isn't consistent with the DSL to be honest!

**Review by ilaif — APPROVED** [review]

**Comment by ilaif** [comment]
LGTM @Midnighter

**Pull request merged**
